include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -113,260 +114,154 @@ struct walk_n
-        (std::is_fundamental_v<oneapi::dpl::__internal::__value_t<_Ranges>> && ...) && __min_type_size < 4;
-    // Vectorize for small types, so we generate 128-byte load / stores in a sub-group
-    constexpr static std::uint8_t __preferred_vector_size =
-        __can_vectorize ? oneapi::dpl::__internal::__dpl_ceiling_div(__max_vector_size, __min_type_size) : 1;
-    constexpr static std::uint8_t __preferred_iters_per_item =
-        __bytes_per_item / (__min_type_size * __preferred_vector_size);
-
-  protected:
-    using __vec_load_t = oneapi::dpl::__par_backend_hetero::__vector_load<__preferred_vector_size>;
-    using __vec_store_t = oneapi::dpl::__par_backend_hetero::__vector_store<__preferred_vector_size>;
-    using __vec_reverse_t = oneapi::dpl::__par_backend_hetero::__vector_reverse<__preferred_vector_size>;
-    using __vec_walk_t = oneapi::dpl::__par_backend_hetero::__vector_walk<__preferred_vector_size>;
-};
-
-// Path that intentionally disables vectorization for algorithms with a scattered access pattern (e.g. binary_search)
-template <typename... _Ranges>
-struct walk_scalar_base
-{
-  private:
-    using _ValueTypes = std::tuple<oneapi::dpl::__internal::__value_t<_Ranges>...>;
-    constexpr static std::uint8_t __min_type_size = oneapi::dpl::__internal::__min_nested_type_size<_ValueTypes>::value;
-    constexpr static std::uint8_t __bytes_per_item = 16;
-
-  public:
-    constexpr static bool __can_vectorize = false;
-    // With no vectorization, the vector size is 1
-    constexpr static std::uint8_t __preferred_vector_size = 1;
-    // To achieve full bandwidth utilization, multiple iterations need to be processed by a work item
-    constexpr static std::uint8_t __preferred_iters_per_item =
-        __bytes_per_item / (__min_type_size * __preferred_vector_size);
-};
-
-template <typename _F, typename _Range>
-struct walk1_vector_or_scalar : public walk_vector_or_scalar_base<_Range>
+// If read accessor returns temporary value then __no_op returns lvalue reference to it.
+// After temporary value destroying it will be a reference on invalid object.
+// So let's don't call functor in case of __no_op
+template <>
+struct walk_n<oneapi::dpl::__internal::__no_op>
 {
-  private:
-    using __base_t = walk_vector_or_scalar_base<_Range>;
-    _F __f;
-    std::size_t __n;
-
-  public:
-    walk1_vector_or_scalar(_F __f, std::size_t __n) : __f(std::move(__f)), __n(__n) {}
-
-    template <typename _IsFull>
-    void
-    __vector_path_impl(_IsFull __is_full, const std::size_t __idx, _Range __rng) const
-    {
-        typename __base_t::__vec_walk_t{__n}(__is_full, __idx, __f, __rng);
-    }
-
-    // _IsFull is ignored here. We assume that boundary checking has been already performed for this index.
-    template <typename _IsFull>
-    void
-    __scalar_path_impl(_IsFull, const std::size_t __idx, _Range __rng) const
-    {
-        __f(__rng[__idx]);
-    }
+    oneapi::dpl::__internal::__no_op __f;
```

Comment:
If there is a reason to have this specialization, and we don't plan to use `__f`, can't we just omit it?

Reply:
same here, not new code...